// @salesforce/source-deploy-retrieve/lib/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.presetMap = exports.getCurrentApiVersion = exports.registry = exports.RegistryAccess = exports.DestructiveChangesType = exports.ComponentSetBuilder = exports.ComponentSet = exports.LazyCollection = exports.ForceIgnore = exports.SourceComponent = exports.ZipTreeContainer = exports.VirtualTreeContainer = exports.NodeFSTreeContainer = exports.TreeContainer = exports.ConnectionResolver = exports.ManifestResolver = exports.MetadataResolver = exports.MetadataConverter = exports.ComponentStatus = exports.RequestStatus = exports.RetrieveResult = exports.MetadataApiRetrieve = exports.DeployResult = exports.MetadataApiDeploy = void 0;
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var client_1 = require("./client");
Object.defineProperty(exports, "MetadataApiDeploy", { enumerable: true, get: function () { return client_1.MetadataApiDeploy; } });
Object.defineProperty(exports, "DeployResult", { enumerable: true, get: function () { return client_1.DeployResult; } });
Object.defineProperty(exports, "MetadataApiRetrieve", { enumerable: true, get: function () { return client_1.MetadataApiRetrieve; } });
Object.defineProperty(exports, "RetrieveResult", { enumerable: true, get: function () { return client_1.RetrieveResult; } });
Object.defineProperty(exports, "RequestStatus", { enumerable: true, get: function () { return client_1.RequestStatus; } });
Object.defineProperty(exports, "ComponentStatus", { enumerable: true, get: function () { return client_1.ComponentStatus; } });
var convert_1 = require("./convert");
Object.defineProperty(exports, "MetadataConverter", { enumerable: true, get: function () { return convert_1.MetadataConverter; } });
var resolve_1 = require("./resolve");
Object.defineProperty(exports, "MetadataResolver", { enumerable: true, get: function () { return resolve_1.MetadataResolver; } });
Object.defineProperty(exports, "ManifestResolver", { enumerable: true, get: function () { return resolve_1.ManifestResolver; } });
Object.defineProperty(exports, "ConnectionResolver", { enumerable: true, get: function () { return resolve_1.ConnectionResolver; } });
Object.defineProperty(exports, "TreeContainer", { enumerable: true, get: function () { return resolve_1.TreeContainer; } });
Object.defineProperty(exports, "NodeFSTreeContainer", { enumerable: true, get: function () { return resolve_1.NodeFSTreeContainer; } });
Object.defineProperty(exports, "VirtualTreeContainer", { enumerable: true, get: function () { return resolve_1.VirtualTreeContainer; } });
Object.defineProperty(exports, "ZipTreeContainer", { enumerable: true, get: function () { return resolve_1.ZipTreeContainer; } });
Object.defineProperty(exports, "SourceComponent", { enumerable: true, get: function () { return resolve_1.SourceComponent; } });
Object.defineProperty(exports, "ForceIgnore", { enumerable: true, get: function () { return resolve_1.ForceIgnore; } });
var collections_1 = require("./collections");
Object.defineProperty(exports, "LazyCollection", { enumerable: true, get: function () { return collections_1.LazyCollection; } });
Object.defineProperty(exports, "ComponentSet", { enumerable: true, get: function () { return collections_1.ComponentSet; } });
Object.defineProperty(exports, "ComponentSetBuilder", { enumerable: true, get: function () { return collections_1.ComponentSetBuilder; } });
Object.defineProperty(exports, "DestructiveChangesType", { enumerable: true, get: function () { return collections_1.DestructiveChangesType; } });
var registry_1 = require("./registry");
Object.defineProperty(exports, "RegistryAccess", { enumerable: true, get: function () { return registry_1.RegistryAccess; } });
Object.defineProperty(exports, "registry", { enumerable: true, get: function () { return registry_1.registry; } });
Object.defineProperty(exports, "getCurrentApiVersion", { enumerable: true, get: function () { return registry_1.getCurrentApiVersion; } });
var presetMap_1 = require("./registry/presets/presetMap");
Object.defineProperty(exports, "presetMap", { enumerable: true, get: function () { return presetMap_1.presetMap; } });
