// @salesforce/source-deploy-retrieve/lib/src/convert/convertContext/decomposedLabelsFinalizer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecomposedLabelsFinalizer = void 0;
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const node_path_1 = require("node:path");
const ts_types_1 = require("@salesforce/ts-types");
const metadata_1 = require("../../utils/metadata");
const constants_1 = require("../../common/constants");
const streams_1 = require("../streams");
const transactionFinalizer_1 = require("./transactionFinalizer");
/**
 * Merges child components that share the same parent in the conversion pipeline
 * into a single file.
 *
 * Inserts unclaimed child components into the parent that belongs to the default package
 */
class DecomposedLabelsFinalizer extends transactionFinalizer_1.ConvertTransactionFinalizer {
    transactionState = {
        customLabelByFullName: new Map(),
    };
    /** to support custom presets (the only way this code should get hit at all pass in the type from a transformer that has registry access */
    customLabelsType;
    // have to maintain the existing interface
    // eslint-disable-next-line @typescript-eslint/require-await, @typescript-eslint/no-unused-vars
    async finalize(defaultDirectory) {
        if (this.transactionState.customLabelByFullName.size === 0) {
            return [];
        }
        return [
            {
                component: {
                    type: (0, ts_types_1.ensure)(this.customLabelsType, 'DecomposedCustomLabelsFinalizer should have set customLabelsType'),
                    fullName: 'CustomLabels',
                },
                writeInfos: [
                    {
                        output: (0, node_path_1.join)((0, ts_types_1.ensure)(this.customLabelsType?.directoryName, 'directoryName missing from customLabels type'), 'CustomLabels.labels'),
                        source: new streams_1.JsToXml(generateXml(this.transactionState.customLabelByFullName)),
                    },
                ],
            },
        ];
    }
}
exports.DecomposedLabelsFinalizer = DecomposedLabelsFinalizer;
/** Return a json object that's built up from the mergeMap children */
const generateXml = (children) => ({
    ['CustomLabels']: {
        [constants_1.XML_NS_KEY]: constants_1.XML_NS_URL,
        // for CustomLabels, that's `labels`
        labels: Array.from(children.values()).filter(metadata_1.customLabelHasFullName).sort(sortLabelsByFullName),
    },
});
const sortLabelsByFullName = (a, b) => a.fullName.localeCompare(b.fullName);
