// @salesforce/source-deploy-retrieve/lib/src/resolve/adapters/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UiBundlesSourceAdapter = exports.DigitalExperienceSourceAdapter = exports.BaseSourceAdapter = exports.DefaultSourceAdapter = exports.DecomposedSourceAdapter = exports.MixedContentSourceAdapter = exports.BundleSourceAdapter = exports.MatchingContentSourceAdapter = void 0;
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var matchingContentSourceAdapter_1 = require("./matchingContentSourceAdapter");
Object.defineProperty(exports, "MatchingContentSourceAdapter", { enumerable: true, get: function () { return matchingContentSourceAdapter_1.MatchingContentSourceAdapter; } });
var bundleSourceAdapter_1 = require("./bundleSourceAdapter");
Object.defineProperty(exports, "BundleSourceAdapter", { enumerable: true, get: function () { return bundleSourceAdapter_1.BundleSourceAdapter; } });
var mixedContentSourceAdapter_1 = require("./mixedContentSourceAdapter");
Object.defineProperty(exports, "MixedContentSourceAdapter", { enumerable: true, get: function () { return mixedContentSourceAdapter_1.MixedContentSourceAdapter; } });
var decomposedSourceAdapter_1 = require("./decomposedSourceAdapter");
Object.defineProperty(exports, "DecomposedSourceAdapter", { enumerable: true, get: function () { return decomposedSourceAdapter_1.DecomposedSourceAdapter; } });
var defaultSourceAdapter_1 = require("./defaultSourceAdapter");
Object.defineProperty(exports, "DefaultSourceAdapter", { enumerable: true, get: function () { return defaultSourceAdapter_1.DefaultSourceAdapter; } });
var baseSourceAdapter_1 = require("./baseSourceAdapter");
Object.defineProperty(exports, "BaseSourceAdapter", { enumerable: true, get: function () { return baseSourceAdapter_1.BaseSourceAdapter; } });
var digitalExperienceSourceAdapter_1 = require("./digitalExperienceSourceAdapter");
Object.defineProperty(exports, "DigitalExperienceSourceAdapter", { enumerable: true, get: function () { return digitalExperienceSourceAdapter_1.DigitalExperienceSourceAdapter; } });
var uiBundlesSourceAdapter_1 = require("./uiBundlesSourceAdapter");
Object.defineProperty(exports, "UiBundlesSourceAdapter", { enumerable: true, get: function () { return uiBundlesSourceAdapter_1.UiBundlesSourceAdapter; } });
