// @salesforce/kit/lib/nodash/internal.js
"use strict";
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEmpty = isEmpty;
exports.lowerFirst = lowerFirst;
exports.camelCaseToTitleCase = camelCaseToTitleCase;
exports.snakeCase = snakeCase;
exports.upperFirst = upperFirst;
exports.toBoolean = toBoolean;
const ts_types_1 = require("@salesforce/ts-types");
/**
 * Checks if value is empty. A value is considered empty unless it’s an arguments object, array, string, or
 * jQuery-like collection with a length greater than 0 or an object with own enumerable properties.
 *
 * @param value The value to inspect.
 */
function isEmpty(value) {
    if (value == null)
        return true;
    if ((0, ts_types_1.isNumber)(value))
        return false;
    if ((0, ts_types_1.isBoolean)(value))
        return false;
    if ((0, ts_types_1.isArrayLike)(value) && value.length > 0)
        return false;
    if ((0, ts_types_1.hasNumber)(value, 'size') && value.size > 0)
        return false;
    if ((0, ts_types_1.isObject)(value) && Object.keys(value).length > 0)
        return false;
    return true;
}
// underlying function
function lowerFirst(value) {
    return value && value.charAt(0).toLowerCase() + value.slice(1);
}
/**
 * Formats a camel case style `string` into a title case.
 *
 * @param text Text to transform.
 */
function camelCaseToTitleCase(text) {
    return text
        .replace(/(^\w|\s\w)/g, (m) => m.toUpperCase())
        .replace(/([A-Z][a-z]+)/g, ' $1')
        .replace(/\s{2,}/g, ' ')
        .trim();
}
// underlying function
function snakeCase(str) {
    return str
        ?.replace(/([a-z])([A-Z])/g, '$1_$2')
        .toLowerCase()
        .replace(/\W/g, '_')
        .replace(/^_+|_+$/g, '');
}
// underlying function
function upperFirst(value) {
    return value && value.charAt(0).toUpperCase() + value.slice(1);
}
/**
 * Converts value to a boolean.
 *
 * @param value The value to convert
 * @returns boolean
 */
function toBoolean(value) {
    switch (typeof value) {
        case 'boolean':
            return value;
        case 'string':
            return value.toLowerCase() === 'true' || value === '1';
        default:
            return false;
    }
}
