// @salesforce/source-deploy-retrieve/lib/src/resolve/forceIgnore.js
"use strict";
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForceIgnore = void 0;
const node_path_1 = require("node:path");
const os = __importStar(require("node:os"));
const index_1 = __importDefault(require("ignore/index"));
const graceful_fs_1 = require("graceful-fs");
const lifecycle_1 = require("@salesforce/core/lifecycle");
const logger_1 = require("@salesforce/core/logger");
const fileSystemHandler_1 = require("../utils/fileSystemHandler");
class ForceIgnore {
    static FILE_NAME = '.forceignore';
    static findCache = new Map();
    static emptySingleton;
    parser;
    forceIgnoreDirectory;
    DEFAULT_IGNORE = [
        '**/*.dup',
        // I know it's ugly.  But I want to be able to retrieve metadata to a local dir, segregated by org.
        // and `.sf` is already ignored in projects, and we already have orgIds for STL
        // so this nastiness is "ignore all dot files except this one directory"
        // once you ignore a parent ex `**/.*` you can't unignore something inside that path, at least with the curent ignore library
        '**/.*',
        '!.sf',
        '**/.sf/**',
        '!**/.sf/orgs',
        '!**/.sf/orgs/**',
        '**/.sf/orgs/*/**',
        '!**/.sf/orgs/*/remoteMetadata',
        '!**/.sf/orgs/*/remoteMetadata/**',
        '**/package2-descriptor.json',
        '**/package2-manifest.json',
    ];
    constructor(forceIgnorePath = '') {
        try {
            const contents = (0, graceful_fs_1.readFileSync)(forceIgnorePath, 'utf-8');
            // check if file `.forceignore` exists
            if (contents !== undefined) {
                // check for windows style separators (\) and warn, that aren't comments
                if (contents.split(os.EOL).find((c) => c.includes('\\') && !c.startsWith('#'))) {
                    // void because you cannot await a method in a constructor
                    void lifecycle_1.Lifecycle.getInstance().emitWarning('Your .forceignore file incorrectly uses the backslash ("\\") as a folder separator; it should use the slash ("/") instead. The ignore rules will not work as expected until you fix this.');
                }
                if (contents.includes('**/unpackaged/**')) {
                    void lifecycle_1.Lifecycle.getInstance().emitWarning('Your .forceignore file contains the "**/unpackaged/**" rule. This will cause all files to be ignored during a retrieve.');
                }
                // add the default ignore paths, and then parse the .forceignore file
                this.parser = (0, index_1.default)().add(`${this.DEFAULT_IGNORE.join('\n')}\n${contents}`);
                this.forceIgnoreDirectory = (0, node_path_1.dirname)(forceIgnorePath);
            }
        }
        catch (e) {
            // TODO: log no force ignore
        }
    }
    /**
     * Performs an upward directory search for a `.forceignore` file and returns a
     * `ForceIgnore` object based on the result. If there is no `.forceignore` file,
     * the returned `ForceIgnore` object will accept everything.
     *
     * Parsed files are cached by absolute path and invalidated when `mtime` or `size` changes on disk.
     *
     * @param seed Path to begin the `.forceignore` search from
     */
    static findAndCreate(seed) {
        const projectConfigPath = (0, fileSystemHandler_1.searchUp)(seed, ForceIgnore.FILE_NAME);
        if (!projectConfigPath) {
            ForceIgnore.emptySingleton ??= new ForceIgnore('');
            return ForceIgnore.emptySingleton;
        }
        const absPath = (0, node_path_1.normalize)(projectConfigPath);
        let mtimeMs;
        let size;
        try {
            const st = (0, graceful_fs_1.statSync)(absPath);
            ({ mtimeMs, size } = st);
        }
        catch {
            return new ForceIgnore('');
        }
        const hit = ForceIgnore.findCache.get(absPath);
        if (hit && hit.mtimeMs === mtimeMs && hit.size === size) {
            return hit.instance;
        }
        const instance = new ForceIgnore((0, node_path_1.join)((0, node_path_1.dirname)(absPath), ForceIgnore.FILE_NAME));
        ForceIgnore.findCache.set(absPath, { mtimeMs, size, instance });
        return instance;
    }
    /** @internal clears module cache; for unit tests only */
    static clearCacheForTest() {
        ForceIgnore.findCache.clear();
        ForceIgnore.emptySingleton = undefined;
    }
    denies(fsPath) {
        if (!this.parser || !this.forceIgnoreDirectory)
            return false;
        try {
            const absoluteFsPath = (0, node_path_1.isAbsolute)(fsPath) ? fsPath : (0, node_path_1.resolve)(this.forceIgnoreDirectory, fsPath);
            const relativePath = (0, node_path_1.relative)(this.forceIgnoreDirectory, absoluteFsPath);
            // node-ignore requires callers to append `/` for directory-only patterns like `node_modules/`:
            // https://github.com/kaelzhang/node-ignore#2-filenames-and-dirnames
            // Known files on disk skip the trailing-slash test to avoid false positives
            // (e.g. a file named `build` matching `build/`).
            // Virtual/zip tree paths that don't exist on disk are assumed to be potential directories.
            const res = this.parser.ignores(relativePath) ||
                (couldBeDirectory(absoluteFsPath) && this.parser.ignores(`${relativePath}/`));
            if (res) {
                logger_1.Logger.childFromRoot('forceIgnore.denies').debug(`Ignoring '${fsPath}' because it matched .forceignore patterns.`);
            }
            return res;
        }
        catch (e) {
            return false;
        }
    }
    accepts(fsPath) {
        return !this.denies(fsPath);
    }
}
exports.ForceIgnore = ForceIgnore;
const couldBeDirectory = (fsPath) => {
    try {
        return !(0, graceful_fs_1.statSync)(fsPath).isFile();
    }
    catch {
        // virtual/zip trees whose paths don't exist on disk — assume it could be a directory
        return true;
    }
};
