// @salesforce/source-deploy-retrieve/lib/src/resolve/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForceIgnore = exports.SourceComponent = exports.ZipTreeContainer = exports.VirtualTreeContainer = exports.NodeFSTreeContainer = exports.TreeContainer = exports.ConnectionResolver = exports.ManifestResolver = exports.MetadataResolver = void 0;
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var metadataResolver_1 = require("./metadataResolver");
Object.defineProperty(exports, "MetadataResolver", { enumerable: true, get: function () { return metadataResolver_1.MetadataResolver; } });
var manifestResolver_1 = require("./manifestResolver");
Object.defineProperty(exports, "ManifestResolver", { enumerable: true, get: function () { return manifestResolver_1.ManifestResolver; } });
var connectionResolver_1 = require("./connectionResolver");
Object.defineProperty(exports, "ConnectionResolver", { enumerable: true, get: function () { return connectionResolver_1.ConnectionResolver; } });
var treeContainers_1 = require("./treeContainers");
Object.defineProperty(exports, "TreeContainer", { enumerable: true, get: function () { return treeContainers_1.TreeContainer; } });
Object.defineProperty(exports, "NodeFSTreeContainer", { enumerable: true, get: function () { return treeContainers_1.NodeFSTreeContainer; } });
Object.defineProperty(exports, "VirtualTreeContainer", { enumerable: true, get: function () { return treeContainers_1.VirtualTreeContainer; } });
Object.defineProperty(exports, "ZipTreeContainer", { enumerable: true, get: function () { return treeContainers_1.ZipTreeContainer; } });
var sourceComponent_1 = require("./sourceComponent");
Object.defineProperty(exports, "SourceComponent", { enumerable: true, get: function () { return sourceComponent_1.SourceComponent; } });
var forceIgnore_1 = require("./forceIgnore");
Object.defineProperty(exports, "ForceIgnore", { enumerable: true, get: function () { return forceIgnore_1.ForceIgnore; } });
