// @salesforce/source-deploy-retrieve/lib/src/resolve/pseudoTypes/agentResolver.js
"use strict";
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseBotVersionFilter = parseBotVersionFilter;
exports.parseBotVersionFullName = parseBotVersionFullName;
exports.resolveAgentMdEntries = resolveAgentMdEntries;
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
const fast_xml_parser_1 = require("fast-xml-parser");
const core_1 = require("@salesforce/core");
const kit_1 = require("@salesforce/kit");
const registry_1 = require("../../registry");
const componentSet_1 = require("../../collections/componentSet");
let logger;
const getLogger = () => {
    if (!logger) {
        logger = core_1.Logger.childFromRoot('resolveAgentMdEntries');
    }
    return logger;
};
/**
 * Parses a bot name to extract version filtering information.
 * Supports patterns:
 * - `BotName_*` - retrieve all versions
 * - `BotName_<number>` - retrieve specific version (e.g., `BotName_3` retrieves version 3)
 * - `BotName` - retrieve highest version
 *
 * @param botName The bot name to parse
 * @returns An object with the base bot name and version filter information
 */
function parseBotVersionFilter(botName) {
    // Handle wildcard pattern: BotName_*
    if (botName.endsWith('_*')) {
        return {
            baseBotName: botName.slice(0, -2),
            versionFilter: 'all',
        };
    }
    // Handle specific version pattern: BotName_<number>
    const versionMatch = botName.match(/^(.+)_(\d+)$/);
    if (versionMatch) {
        const [, baseName, versionStr] = versionMatch;
        const versionNum = parseInt(versionStr, 10);
        return {
            baseBotName: baseName,
            versionFilter: versionNum,
        };
    }
    // Default: retrieve highest version
    return {
        baseBotName: botName,
        versionFilter: 'highest',
    };
}
/**
 * Parses a BotVersion fullName from a manifest or metadata entry to extract bot name and version.
 * Supports patterns:
 * - `BotName.vN` - specific version (e.g., `Local_Info_Agent.v4`)
 * - `BotName.*` - all versions
 * - `BotName.versionPart` - unknown format, extracts bot name and defaults to highest
 * - `BotName` - no version specified, defaults to highest
 *
 * @param fullName The BotVersion fullName from manifest/metadata
 * @returns An object with the bot name and version filter
 */
function parseBotVersionFullName(fullName) {
    // Handle standard BotVersion format: BotName.vN
    const match = fullName.match(/^(.+)\.v(\d+)$/);
    if (match) {
        const [, botName, versionStr] = match;
        const versionNum = parseInt(versionStr, 10);
        return { botName, versionFilter: versionNum };
    }
    // Handle BotName.* pattern (all versions)
    if (fullName.endsWith('.*')) {
        return {
            botName: fullName.substring(0, fullName.length - 2),
            versionFilter: 'all',
        };
    }
    // Handle unknown formats with a dot
    const dotIndex = fullName.lastIndexOf('.');
    if (dotIndex > 0) {
        const botName = fullName.substring(0, dotIndex);
        return { botName, versionFilter: 'highest' };
    }
    // No dot in name, treat entire string as bot name
    return { botName: fullName, versionFilter: 'highest' };
}
/**
 * This is the local "spidering" logic for agents.  Given the API name for a Bot,
 * and either an org connection or local file system paths, resolve to the top
 * level agent metadata. E.g., Bot, BotVersion, GenAiPlanner, GenAiPlugin, and
 * GenAiFunction.
 *
 * If an org connection is provided, it will query the org for GenAiPlanner and
 * GenAiPlugin metadata associated with the Bot name.
 *
 * If no org connection but directory paths are provided, it will search those
 * directory paths for BotVersion and GenAiPlanner metadata associated with the
 * Bot name.
 *
 * @param agentMdInfo Data necessary to get agent related metadata.
 * @returns An array of metadata types and possibly metadata names (Metadata entries)
 */
async function resolveAgentMdEntries(agentMdInfo) {
    const { botName, connection, directoryPaths } = agentMdInfo;
    const v63topLevelMd = ['Bot', 'GenAiPlanner', 'GenAiPlugin', 'GenAiFunction'];
    const v64topLevelMd = ['Bot', 'GenAiPlannerBundle', 'GenAiPlugin', 'GenAiFunction'];
    let debugMsg = `Resolving agent metadata with botName: ${botName}`;
    if (connection) {
        debugMsg += ` and org connection ${connection.getUsername()}`;
    }
    if (directoryPaths) {
        debugMsg += ` in paths: ${directoryPaths.join(', ')}`;
    }
    getLogger().debug(debugMsg);
    if (botName === '*') {
        // Get all Agent top level metadata
        if (connection && Number(connection.getApiVersion()) > 63.0) {
            return Promise.resolve(v64topLevelMd);
        }
        return Promise.resolve(v63topLevelMd);
    }
    if (connection) {
        return resolveAgentFromConnection(connection, botName);
    }
    else {
        if (!directoryPaths || directoryPaths?.length === 0) {
            throw core_1.SfError.create({
                message: 'Cannot resolve Agent pseudo type from local files without a source directory',
            });
        }
        const registry = agentMdInfo.registry ?? new registry_1.RegistryAccess();
        return resolveAgentFromLocalMetadata(botName, directoryPaths, registry);
    }
}
// Queries the org for metadata related to the provided Bot API name and returns those
// metadata type:name pairs.
const resolveAgentFromConnection = async (connection, botName) => {
    const mdEntries = [`Bot:${botName}`];
    // Query the org for agent metadata related to the Bot API name.
    try {
        // Query for the GenAiPlannerId
        const genAiPlannerIdQuery = `SELECT Id FROM GenAiPlannerDefinition WHERE DeveloperName = '${botName}'`;
        const plannerId = (await connection.singleRecordQuery(genAiPlannerIdQuery, { tooling: true })).Id;
        if (plannerId) {
            const plannerType = Number(connection.getApiVersion()) > 63.0 ? 'GenAiPlannerBundle' : 'GenAiPlanner';
            mdEntries.push(`${plannerType}:${botName}`);
            // Query the junction table to get all GenAiPlugins associated with this GenAiPlannerId
            const topicIdsByPlanner = (await connection.tooling.query(`SELECT Plugin FROM GenAiPlannerFunctionDef WHERE PlannerId = '${plannerId}'`)).records;
            if (topicIdsByPlanner.length) {
                // Query the GenAiPluginDefinition table to get the DeveloperName for each GenAiPluginId
                const genAiPluginIds = `'${topicIdsByPlanner
                    .map((record) => record.Plugin)
                    .filter((pluginId) => (0, core_1.validateSalesforceId)(pluginId))
                    .join("','")}'`;
                // This query returns customized plugins only
                const genAiPluginNames = (await connection.tooling.query(`SELECT DeveloperName FROM GenAiPluginDefinition WHERE Id IN (${genAiPluginIds})`)).records;
                if (genAiPluginNames.length) {
                    genAiPluginNames.map((r) => mdEntries.push(`GenAiPlugin:${r.DeveloperName}`));
                }
                else {
                    getLogger().debug(`No GenAiPlugin metadata matches for plannerId: ${plannerId}. Reading the ${plannerType} metadata for plugins...`);
                    await getPluginNamesFromPlanner(connection, botName, mdEntries);
                }
            }
            else {
                getLogger().debug(`No GenAiPlugin metadata matches for plannerId: ${plannerId}. Reading the ${plannerType} metadata for plugins...`);
                await getPluginNamesFromPlanner(connection, botName, mdEntries);
            }
        }
        else {
            getLogger().debug(`No GenAiPlanner metadata matches for Bot: ${botName}`);
        }
    }
    catch (err) {
        const wrappedErr = core_1.SfError.wrap(err);
        getLogger().debug(`Error when querying for GenAiPlanner or GenAiPlugin by Bot name: ${botName}\n${wrappedErr.message}`);
        if (wrappedErr.stack) {
            getLogger().debug(wrappedErr.stack);
        }
    }
    return mdEntries;
};
// Finds and reads local metadata files related to the provided Bot API name and
// returns those metadata type:name pairs.
const resolveAgentFromLocalMetadata = (botName, directoryPaths, registry) => {
    const mdEntries = new Set([`Bot:${botName}`]);
    // Inspect local files for agent metadata related to the Bot API name
    const botType = registry.getTypeByName('Bot');
    const botCompSet = componentSet_1.ComponentSet.fromSource({
        fsPaths: directoryPaths,
        include: new componentSet_1.ComponentSet([{ type: botType, fullName: botName }], registry),
        registry,
    });
    if (botCompSet.size < 1) {
        getLogger().debug(`Cannot resolve botName: ${botName} to a local file`);
    }
    const parser = new fast_xml_parser_1.XMLParser({ ignoreAttributes: false });
    const botFiles = botCompSet.getComponentFilenamesByNameAndType({ type: 'Bot', fullName: botName });
    let plannerType = registry.getTypeByName('GenAiPlannerBundle');
    let plannerTypeName = 'GenAiPlannerBundle';
    let plannerCompSet = componentSet_1.ComponentSet.fromSource({
        fsPaths: directoryPaths,
        include: new componentSet_1.ComponentSet([{ type: plannerType, fullName: botName }], registry),
        registry,
    });
    // If the plannerCompSet is empty it might be due to the GenAiPlannerBundle having a
    // different name than the Bot. We need to search the BotVersion for the
    // planner API name.
    if (plannerCompSet.size < 1) {
        const botVersionFile = botFiles.find((botFile) => botFile.endsWith('.botVersion-meta.xml'));
        if (botVersionFile) {
            getLogger().debug(`Reading and parsing ${botVersionFile} to find all GenAiPlanner/GenAiPlannerBundle references`);
            const botVersionJson = xmlToJson(botVersionFile, parser);
            // Per the schema, there can be multiple GenAiPlanners linked to a BotVersion
            // but I'm not sure how that would work so for now just using the first one.
            const planners = (0, kit_1.ensureArray)(botVersionJson.BotVersion.conversationDefinitionPlanners);
            const genAiPlannerName = planners.length ? planners[0]?.genAiPlannerName : undefined;
            if (genAiPlannerName) {
                plannerCompSet = componentSet_1.ComponentSet.fromSource({
                    fsPaths: directoryPaths,
                    include: new componentSet_1.ComponentSet([{ type: plannerType, fullName: genAiPlannerName }], registry),
                    registry,
                });
                // If the plannerCompSet is empty look for a GenAiPlanner
                if (plannerCompSet.size < 1) {
                    plannerTypeName = 'GenAiPlanner';
                    plannerType = registry.getTypeByName('GenAiPlanner');
                    plannerCompSet = componentSet_1.ComponentSet.fromSource({
                        fsPaths: directoryPaths,
                        include: new componentSet_1.ComponentSet([{ type: plannerType, fullName: genAiPlannerName }], registry),
                        registry,
                    });
                    if (plannerCompSet.size < 1) {
                        getLogger().debug(`Cannot find GenAiPlanner or GenAiPlannerBundle with name: ${genAiPlannerName}`);
                    }
                }
                getLogger().debug(`Adding ${plannerTypeName}:${genAiPlannerName}`);
                mdEntries.add(`${plannerTypeName}:${genAiPlannerName}`);
            }
            else {
                getLogger().debug(`Cannot find GenAiPlannerName in BotVersion file: ${botVersionFile}`);
            }
        }
    }
    else {
        getLogger().debug(`Adding ${plannerTypeName}:${botName}`);
        mdEntries.add(`${plannerTypeName}:${botName}`);
    }
    // Read the GenAiPlanner or GenAiPlannerBundle file for GenAiPlugins
    const plannerComp = plannerCompSet.getSourceComponents().first();
    if (plannerComp) {
        let plannerFilePath;
        if (plannerTypeName === 'GenAiPlannerBundle' && plannerComp.content) {
            const plannerFileName = plannerComp.tree
                .readDirectory(plannerComp.content)
                .find((p) => p.endsWith('.genAiPlannerBundle'));
            if (plannerFileName) {
                plannerFilePath = (0, node_path_1.join)(plannerComp.content, plannerFileName);
            }
            else {
                getLogger().debug(`Cannot find GenAiPlannerBundle file in ${plannerComp.content}`);
            }
        }
        else {
            plannerFilePath = plannerComp.xml;
        }
        // Certain internal plugins and functions cannot be retrieved/deployed so don't include them.
        const internalPrefixes = ['EmployeeCopilot__', 'SvcCopilotTmpl__'];
        if (plannerFilePath) {
            getLogger().debug(`Reading and parsing ${plannerFilePath} to find all GenAiPlugin references`);
            const plannerJson = xmlToJson(plannerFilePath, parser);
            // Add plugins defined in the planner
            // @ts-expect-error temporary until GenAiPlannerBundle metadata type is defined
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            const genAiPlugins = (0, kit_1.ensureArray)(plannerJson[plannerTypeName].genAiPlugins);
            const pluginType = registry.getTypeByName('GenAiPlugin');
            const genAiPluginComps = [];
            genAiPlugins?.map((plugin) => {
                if (plugin.genAiPluginName && !internalPrefixes.some((prefix) => plugin.genAiPluginName?.startsWith(prefix))) {
                    genAiPluginComps.push({ type: pluginType, fullName: plugin.genAiPluginName });
                    getLogger().debug(`Adding GenAiPlugin:${plugin.genAiPluginName}`);
                    mdEntries.add(`GenAiPlugin:${plugin.genAiPluginName}`);
                }
            });
            // Add functions defined in the plugins
            if (genAiPluginComps.length) {
                const pluginCompSet = componentSet_1.ComponentSet.fromSource({
                    fsPaths: directoryPaths,
                    include: new componentSet_1.ComponentSet(genAiPluginComps, registry),
                    registry,
                });
                if (pluginCompSet.size > 1) {
                    // For all plugin files, read and parse, adding all functions.
                    for (const comp of pluginCompSet.getSourceComponents()) {
                        if (comp.xml) {
                            getLogger().debug(`Reading and parsing ${comp.xml} to find all GenAiFunction references`);
                            const genAiPlugin = xmlToJson(comp.xml, parser);
                            const genAiFunctions = (0, kit_1.ensureArray)(genAiPlugin.GenAiPlugin.genAiFunctions);
                            genAiFunctions.map((func) => {
                                if (func.functionName && !internalPrefixes.some((prefix) => func.functionName.startsWith(prefix))) {
                                    getLogger().debug(`Adding GenAiFunction:${func.functionName}`);
                                    mdEntries.add(`GenAiFunction:${func.functionName}`);
                                }
                            });
                        }
                    }
                }
            }
        }
    }
    return Array.from(mdEntries);
};
// Read an xml file, parse it to json and return the JSON.
const xmlToJson = (path, parser) => {
    const file = (0, node_fs_1.readFileSync)(path, 'utf8');
    if (!file)
        throw new core_1.SfError(`No metadata file found at ${path}`);
    return parser.parse(file);
};
// read the GenAiPlanner or GenAiPlannerBundle metadata from the org and add the GenAiPlugin names to the mdEntries array
const getPluginNamesFromPlanner = async (connection, botName, mdEntries) => {
    const plannerType = Number(connection.getApiVersion()) > 63.0 ? 'GenAiPlannerBundle' : 'GenAiPlanner';
    // @ts-expect-error jsForce types don't know about GenAiPlanner yet
    const genAiPlannerMd = await connection.metadata.read(plannerType, botName);
    const genAiPlannerMdArr = (0, kit_1.ensureArray)(genAiPlannerMd);
    if (genAiPlannerMdArr?.length && genAiPlannerMdArr[0]?.genAiPlugins.length) {
        genAiPlannerMdArr[0].genAiPlugins.map((plugin) => {
            if (plugin.genAiPluginName?.length) {
                mdEntries.push(`GenAiPlugin:${plugin.genAiPluginName}`);
            }
        });
    }
    else {
        getLogger().debug(`No GenAiPlugin metadata found in planner file for API name: ${botName}`);
    }
};
