// @salesforce/source-deploy-retrieve/lib/src/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.baseWithoutSuffixes = exports.parseNestedFullName = exports.trimUntil = exports.parentName = exports.parseMetadataXml = exports.baseName = exports.extName = exports.trimMetaXmlSuffix = exports.generateMetaXMLPath = exports.generateMetaXML = void 0;
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var metadata_1 = require("./metadata");
Object.defineProperty(exports, "generateMetaXML", { enumerable: true, get: function () { return metadata_1.generateMetaXML; } });
Object.defineProperty(exports, "generateMetaXMLPath", { enumerable: true, get: function () { return metadata_1.generateMetaXMLPath; } });
Object.defineProperty(exports, "trimMetaXmlSuffix", { enumerable: true, get: function () { return metadata_1.trimMetaXmlSuffix; } });
var path_1 = require("./path");
Object.defineProperty(exports, "extName", { enumerable: true, get: function () { return path_1.extName; } });
Object.defineProperty(exports, "baseName", { enumerable: true, get: function () { return path_1.baseName; } });
Object.defineProperty(exports, "parseMetadataXml", { enumerable: true, get: function () { return path_1.parseMetadataXml; } });
Object.defineProperty(exports, "parentName", { enumerable: true, get: function () { return path_1.parentName; } });
Object.defineProperty(exports, "trimUntil", { enumerable: true, get: function () { return path_1.trimUntil; } });
Object.defineProperty(exports, "parseNestedFullName", { enumerable: true, get: function () { return path_1.parseNestedFullName; } });
Object.defineProperty(exports, "baseWithoutSuffixes", { enumerable: true, get: function () { return path_1.baseWithoutSuffixes; } });
