// @salesforce/source-deploy-retrieve/lib/src/client/index.js
"use strict";
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentStatus = exports.RequestStatus = exports.RetrieveResult = exports.MetadataApiRetrieve = exports.DeployResult = exports.MetadataApiDeploy = void 0;
var metadataApiDeploy_1 = require("./metadataApiDeploy");
Object.defineProperty(exports, "MetadataApiDeploy", { enumerable: true, get: function () { return metadataApiDeploy_1.MetadataApiDeploy; } });
Object.defineProperty(exports, "DeployResult", { enumerable: true, get: function () { return metadataApiDeploy_1.DeployResult; } });
var metadataApiRetrieve_1 = require("./metadataApiRetrieve");
Object.defineProperty(exports, "MetadataApiRetrieve", { enumerable: true, get: function () { return metadataApiRetrieve_1.MetadataApiRetrieve; } });
Object.defineProperty(exports, "RetrieveResult", { enumerable: true, get: function () { return metadataApiRetrieve_1.RetrieveResult; } });
var types_1 = require("./types");
Object.defineProperty(exports, "RequestStatus", { enumerable: true, get: function () { return types_1.RequestStatus; } });
Object.defineProperty(exports, "ComponentStatus", { enumerable: true, get: function () { return types_1.ComponentStatus; } });
