// @salesforce/source-deploy-retrieve/lib/src/common/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XML_COMMENT_PROP_NAME = exports.XML_NS_KEY = exports.XML_NS_URL = exports.XML_DECL = exports.META_XML_SUFFIX = exports.DEFAULT_PACKAGE_ROOT_SFDX = void 0;
/*
 * Copyright 2026, Salesforce, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const node_path_1 = require("node:path");
exports.DEFAULT_PACKAGE_ROOT_SFDX = (0, node_path_1.join)('main', 'default');
exports.META_XML_SUFFIX = '-meta.xml';
exports.XML_DECL = '<?xml version="1.0" encoding="UTF-8"?>\n';
exports.XML_NS_URL = 'http://soap.sforce.com/2006/04/metadata';
exports.XML_NS_KEY = '@_xmlns';
exports.XML_COMMENT_PROP_NAME = '#xml__comment';
